File: day12/sol.go
package main

import (
	"fmt"
	"os"
	"strings"
	"time"
)

type Record struct {
	record string
	groups []int
}

type State struct {
	r_i, g_i, streak int
}

func ReadFile(filename string) string {
	bytes, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return strings.TrimSpace(string(bytes))
}

func ParseInt(s string) int {
	n := 0
	for _, c := range s {
		n = n*10 + int(c-'0')
	}
	return n
}

func ParseInput(text string) []Record {
	lines := strings.Split(text, "\n")
	records := make([]Record, len(lines))
	for i, line := range lines {
		fields := strings.Split(line, " ")

		record := fields[0]
		groups_str := strings.Split(fields[1], ",")
		groups := make([]int, len(groups_str))
		for j, group_str := range groups_str {
			groups[j] = ParseInt(group_str)
		}

		records[i] = Record{record, groups}
	}
	return records
}

func Solve(record Record, cache map[State]int, r_i, g_i, streak int) int {
	state := State{r_i, g_i, streak}
	if val, ok := cache[state]; ok {
		return val
	}

	if r_i >= len(record.record) {
		if streak == 0 && g_i == len(record.groups) {
			return 1
		}
		if g_i == len(record.groups)-1 && streak == record.groups[g_i] {
			return 1
		}
		return 0
	}

	n := 0
	switch record.record[r_i] {
	case '.':
		if streak == 0 {
			n = Solve(record, cache, r_i+1, g_i, 0)
		} else if g_i < len(record.groups) && streak == record.groups[g_i] {
			n = Solve(record, cache, r_i+1, g_i+1, 0)
		}
	case '#':
		n = Solve(record, cache, r_i+1, g_i, streak+1)
	case '?':
		if streak == 0 {
			n = Solve(record, cache, r_i+1, g_i, 0)
		} else if g_i < len(record.groups) && streak == record.groups[g_i] {
			n = Solve(record, cache, r_i+1, g_i+1, 0)
		}
		n += Solve(record, cache, r_i+1, g_i, streak+1)
	}

	cache[state] = n
	return n
}

func Part1(records []Record) int {
	sum := 0
	for _, record := range records {
		sum += Solve(record, make(map[State]int), 0, 0, 0)
	}
	return sum
}

func Part2(records []Record) int {
	sum := 0
	for _, record := range records {
		r := make([]byte, len(record.record)*5)
		for i := 0; i < 5; i++ {
			copy(r[i*len(record.record):(i+1)*len(record.record)], record.record)
		}
		g := make([]int, len(record.groups)*5)
		for i := 0; i < 5; i++ {
			copy(g[i*len(record.groups):(i+1)*len(record.groups)], record.groups)
		}

		sum += Solve(Record{string(r), g}, make(map[State]int), 0, 0, 0)
	}
	return sum
}
func PrintTime(start time.Time, msg string) {
	dur := float64(time.Since(start).Nanoseconds())
	switch {
	case dur < 1e3:
		fmt.Printf("%v %.3fns\n", msg, dur)
	case dur < 1e6:
		fmt.Printf("%v %.3fµs\n", msg, dur/1e3)
	case dur < 1e9:
		fmt.Printf("%v %.3fms\n", msg, dur/1e6)
	default:
		fmt.Printf("%v %.3fs\n", msg, dur/1e9)
	}
}

func main() {
	start := time.Now()
	defer PrintTime(start, "Elapsed time:")

	text := ReadFile("inputs/day12.txt")
	records := ParseInput(text)

	fmt.Printf("Part1: %d\n", Part1(records))
	fmt.Printf("Part2: %d\n", Part2(records))
}

File: day12/sol.py
from timing_util import Timing


def get_data(content: str) -> list[tuple[str, list[int]]]:
    data = []
    for line in content.splitlines():
        record, groups = line.split()
        groups = list(map(int, groups.split(",")))
        data.append((record, groups))

    return data


def sol(record: str, groups: list[int]) -> int:
    def solve(r_i: int, g_i: int, streak: int) -> int:
        if (state := (r_i, g_i, streak)) in cache:
            # if superpositions converge, we only need to compute one of them
            return cache[state]

        if r_i >= R:
            # record is exhausted and we have consumed all groups
            if not streak and g_i == G:
                return 1
            # we finished on a complete streak
            elif g_i == G - 1 and streak == groups[g_i]:
                return 1
            # superposition did not work -> no solution was found
            else:
                return 0

        n = 0
        match record[r_i]:
            case ".":
                if not streak:
                    # no streak before, just advance one in record
                    n = solve(r_i + 1, g_i, 0)
                elif g_i < G and groups[g_i] == streak:
                    # if current streak of '#' is over, advance one in record and arrangements
                    n = solve(r_i + 1, g_i + 1, 0)
            case "#":
                # advance one in record and current streak
                n = solve(r_i + 1, g_i, streak + 1)
            case "?":
                # for '?' we superposition the two cases above
                if not streak:
                    n = solve(r_i + 1, g_i, 0)
                elif g_i < G and groups[g_i] == streak:
                    n = solve(r_i + 1, g_i + 1, 0)

                n += solve(r_i + 1, g_i, streak + 1)

        cache[state] = n
        return n

    R, G = len(record), len(groups)
    cache = {}
    return solve(r_i=0, g_i=0, streak=0)


def part1(data: list[tuple[str, list[int]]]) -> int:
    return sum(sol(record, groups) for record, groups in data)


def part2(data: list[tuple[str, list[int]]]) -> int:
    return sum(sol("?".join([record] * 5), groups * 5) for record, groups in data)


if __name__ == "__main__":
    with Timing("Elapsed time: "):
        with open("inputs/day12.txt") as in_file:
            data = get_data(in_file.read().strip())

        print(f"part1: {part1(data)}")
        print(f"part2: {part2(data)}")
